lib/flux/app-state.js

```diff
@@ -452,6 +417,28 @@ export const actionMap = new ActionMap({
       },
     },
 
+    /**
+     * A note is being changed from somewhere else! If the same
+     * note is also open and being edited, we need to make sure
+     * any in-memory changes don't get blown away. This is our
+     * chance to tell node-simperium what we want the note to
+     * be.
+     *
+     * node-simperium will compare these changes with the changes
+     * from the server and merge them together.
+     */
+    onNoteBeforeRemoteUpdate: {
+      creator({ noteId }) {
+        return (dispatch, getState) => {
+          const state = getState();
+          if (state.selectedNoteId !== noteId) {
+            return null;
+          }
+          return getState().note.data;
```

Comment:
sadly we need more changes to the app because we're trapping content inside the `debounce` of `<NoteDetail />`

what's _encouraging_ is that I think we can completely get rid of the debounce inside the editor and move it here or somewhere else. what can happen while these async things are floating in lala land is that we can receive a remove update from the network, send a copy of our note from the app memory to simperium via `onNoteBeforeRemoteUpdate`, get the `update` from `node-simperium` and update the editor contents…_and then_ as we update the editor and _after_ simperium has finished doing its thing the editor says, "oh hey, I have changes waiting to send, I'll push them out now." this cycles back up the app and flushes to Simperium and ends up wiping out the new changes

```
= START =
simperium: AC
app: AC
editorDebounce: [empty]

= TYPE =
simperium: AC
app: AC
editorDebounce: ACD

= NETWORK UPDATE =
simperium: ABC
app: AC
editorDebounce: ACD

= BEFORE NETWORK CHANGE =
app: "I have 'AC'"
simperium: "Okay, no changes, apply update and we now have ABC"
editorDebounce: ACD

= UPDATE =
simperium: ABC "hey app, this is the new version"
app: ABC "okay, thanks, I'll replace the text in my editor"
editorDebounce: ACD

= UPDATE EDITOR COMPONENT PROPS =
app: ABC "hey I have changes, I should flush them"

= DEBOUNCE FLUSH =
simperium: ACD [-B]
app: ABC
editorDebounce: [empty]

= SIMPERIUM RESPONDS =
simperium: ACD
app: ACD
editorDebounce: [empty]
```

I predict this cycle. This patch is probably _better_ than what we have, but I think we need to make sure we can flush the editor debounce during this `subscribe` otherwise we'll immediately undo the update from the network

Reply:
Ideally the debounce would be related to UI flushes while the state can be manipulated freely (isn't React suspense supposed to handle this in the future anyway).

Reply:
_If_ we remove the debounce from the editor itself then it should mirror the app's note state and we can think of it as always being immediately in sync.

_Then_ if we go down ten lines in `updateNoteContent` (which will now be called on every keypress or note content change) we can wrap `noteBucket.update` in the debounce. We'd have to expose that debounce somehow to `onNoteBeforeRemoteUpdate` so it can cancel it though.

this would have to happen _per note_ and _might_ be reasonable to share via `app.jsx`.

```js
// app.jsx
class NoteChangeQueue {
	constructor( noteBucket ) {
		this.queue = new Map();
		this.noteBucket = noteBucket;
	}

	add( noteId, data ) {
    this.remove( noteId );

		const runner = setTimeout( () => {
			const [ hasQueue, noteData ] = this.remove( noteId );
			if ( hasQueue ) {
				this.noteBucket.update( noteId, noteData );
			}
		}, 2000 );

		this.queue.add( noteId, [ runner, data ] );
  }

	remove( noteId ) {
		const hasQueue = this.queue.has( noteId );
		const [ runner, noteData ] = hasQueue ? this.queue.get( noteId ) : [ null, null ];

		if ( runner ) {
			clearTimeout( runner );
		}

		this.queue.delete( noteId );
		return [ hasQueue, noteData ];
	}
}

…

this.queuedNotes = new NoteChangeQueue( this.props.noteBucket );

this.props.noteBucket.on(…).….subscribe( noteId => {
	this.props.actions.onNoteBeforeRemoteUpdate({
		noteId,
		queuedNotes: this.queuedNotes,
	})
} )

…

onUpdateContent = ( note, content ) =>
	this.props.actions.updateNoteContent( {
		noteBucket: this.props.noteBucket,
		note,
		content,
		queuedNotes: this.queuedNotes
	} );


// app-state.js
onNoteBeforeRemoteUpdate: {
	creator({ noteId, queuedNotes }) {
		return (dispatch, getState) => {
			…

			const [ wasQueued, noteData ] = queuedNotes.remove( noteId );

			// if we have changes we haven't yet sent to Simperium then return
			// them as the new `local` value, but otherwise use the existing app state
			return wasQueued ? noteData : getState().note.data;
		}
	} 
}

updateNoteContent: {
	creator({ noteBucket, note, content, queuedNotes }) {
		return (dispatch, getState) => {
			…

			// update the bucket and sync
			// noteBucket.update(note.id, note.data); <<< no longer this
			// instead we queue up the changes and update app state
			queuedNotes.add( note.id, note.data );
			dispatch(update(getState(), {
				note: { $set: note.data }
			})
			…
		};
	}
}
```

Reply:
> Ideally the debounce would be related to UI flushes while the state can be manipulated freely (isn't React suspense supposed to handle this in the future anyway).

Maybe, but we still need to rearrange where the debounce happens and what it does else we'll repeat the bug I have above.

```js
// app-state.jsx
updateNoteContent: {
	creator({ noteBucket, note, content }) {
		return (dispatch, getState) => {
			// immediately update app state
			dispatch(update(getState(), {note: {$set: {...note.data,content}}}))
		}
	}
}

// because this is just notifying the `node-simperium` library we don't have to
// do anything. our app state is already updated and won't change just because
// we submit the changes. we only care about updates through `.on( 'update' )`
const sendNoteToSimperium = note => noteBucket.update( note.id, note.data );

// note-detail/index.js
saveNote = content => {
	const { note } = this.props;

	if ( ! this.isValidNote( note ) ) {
		return;
	}

	this.props.onNoteChange( note, content ); // immediately update app state
	this.queueUpdate( note.id );
}

queueUpdate = debounce( () => {
	const { note } = this.props;

	// tell the app to run noteBucket.update
	this.props.sendNoteToSimperium( note );
}, 2000 )

componentWillUnmount() {
	// when closing this note make sure we get our updates to Simperium
	this.queueUpdate.flush();
}
```

this way I think would require fewer changes and fewer new concepts - a new updater function in `app-state` that gets passed down to the editor component - and may be closer to the way the mobile libraries work.

Reply:
> we can wrap noteBucket.update in the debounce. 

This sounds like the correct thing. If it's sending the data to simperium that we want to debounce then debouncing the state was an indirect way of doing that.

> We'd have to expose that debounce somehow to onNoteBeforeRemoteUpdate so it can cancel it though.

Why de we need to be able to cancel it? Calling extra `noteBucket.updates` isn't a problem.


```js
			// if we have changes we haven't yet sent to Simperium then return
			// them as the new `local` value, but otherwise use the existing app state
			return wasQueued ? noteData : getState().note.data;
```

I don't think we need to do this. If we return `false`y from here, the default implementation will use what is in the `noteBucket` as the "current state". If a note isn't open it doesn't have a in-memory representation we need to worry about.

Reply:
This is the debounce we're concerned with correct?

https://github.com/Automattic/simplenote-electron/blob/31e6bc7b2dfaae2059bbee8499308eaaf1898547/lib/note-detail/index.jsx#L44

Reply:
**This** is where we should "debounce". And _not_ debounce the `update` function itself but  use the `sync: true|false` to reduce sync operations, because that is the point of the debounce right?

https://github.com/Automattic/simplenote-electron/blob/31e6bc7b2dfaae2059bbee8499308eaaf1898547/lib/flux/app-state.js#L463

Reply:
all this sounds good. _if_ we inform `node-simperium` of our changes immediately upon making them then there will never be a time when we have a different copy of the local data and therefore no need to send an updated copy of the local changes `beforeNetworkChange`. well, that's true _mostly_ but while we still have indexedDB in the picture we have very small slivers of time when we could have multiple copies - that's nothing to worry about here.